Generate switches for harmony activities automatically

mkeesey opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This adds switches for each activity associated with a Harmony remote.

Harmony remotes have the concept of "Activities" which represent a set of actions like "Watch Fire TV" which will be configured to turn on your television, turn on your receiver, and turn on the Fire TV stick. Currently the harmony integration supports triggering these activities through a service call to the `remote.turn_on` service with the activity name as a property of the call. To set up switches to control each activity today, the typical advice is to set up template switches to call these services and to extract the current_activity from the remote to populate the value_template.

This PR removes the need to manually set up these template switches, and instead adds a switch per activity. These switches, by default, are named `{hub name} {activity name}` (e.g. "Guest Room Watch TV") to support multiple hubs. To support testing these changes, I also enhanced the mock code in tests to be able to keep state so we can more effectively test state changes.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

No current documentation update. From integrations I've used before I don't believe they tend to document all of the entities they expose. However, happy to update docs in whatever way makes sense to you folks.

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mkeesey committed:
Adding switch code for harmony activities

mkeesey committed:
Working on-off

mkeesey committed:
Removing poll code for now

mkeesey committed:
Async updates for current activity

mkeesey committed:
Update our state based on events

mkeesey committed:
Notifications we got connected or disconnected

mkeesey committed:
Remove unncessary constructor arg

mkeesey committed:
Initial switch tests

mkeesey committed:
Additional tests for switch transitions

mkeesey committed:
Test transitions for availability

mkeesey committed:
Testing switch state changes

mkeesey committed:
Tests passing

mkeesey committed:
Final tests

mkeesey committed:
Updating manifest.

homeassistant commented:
Hi @mkeesey, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42331).

Once you do that we will be able to review and accept this pull request.

Thanks!

mkeesey committed:
Correctly mock the return value from a call to the library

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -311,6 +316,16 @@ def new_activity(self, activity_info: tuple) -> None:
         self._state = bool(activity_id != -1)
         self._available = True
         self.async_write_ha_state()
+        async_dispatcher_send(
```
We shouldn't dispatch from an entity.

`self._client = HarmonyClient(ip_address=host)` needs to be moved out of the entity if we want to use it for multiple entities.

We can setup subscriber callbacks in the entities that want to consume the callsbacks from it.

An example is in `august/subscriber.py`

mkeesey replied:
Really helpful example, thank you! I've moved the HarmonyClient to a centralized data object which allows entities to subscribe for callbacks and centralizes some of the logic for common operations (like changing activities).

bdraco commented on homeassistant/components/harmony/switch.py:
```diff
@@ -0,0 +1,110 @@
+
+    async def async_added_to_hass(self):
+        """Call when entity is added to hass."""
+        self._dispatcher_disconnectors.append(
```
These can be wrapped in `self.async_on_remove` instead.

However the integration needs to be refactored to move `HarmonyClient` out of the remote entity before it should be used for other entities.

bdraco reviewed (changes requested):
There is some refactoring work that needs to be done to avoid one entity depending on another since entities can be disabled.

Please see comments above.

mkeesey committed:
Adding new subscriber classes

mkeesey committed:
Update class name and location

mkeesey committed:
Got the refactor working locally.

mkeesey committed:
Tests passing

mkeesey committed:
Tracking state changes

mkeesey committed:
Remove write_to_config_file - this appears to never be read.

It was added far back in the past to account for a harmony library
change, but nothing ever reads that path.

Removing that side effect from tests is a pain - avoid the side effect
completely.

mkeesey committed:
Connection changes tested

mkeesey committed:
Clean up temporary code

mkeesey committed:
Update .coveragerc for harmony component

Specifically exclude untested files instead of the whole module

mkeesey commented on homeassistant/components/harmony/data.py:
```diff
@@ -0,0 +1,333 @@
+    async def async_start_activity(self, activity: str):
+        """Start an activity from the Harmony device."""
+
+        if activity:
```
Many of these long methods for starting activities, sending commands, etc were directly moved from the remote code with very minor modifications to support their new location.

mkeesey commented on homeassistant/components/harmony/remote.py:
```diff
@@ -405,94 +319,21 @@ async def async_send_command(self, command, **kwargs):
-        else:
-            await self.hass.async_add_executor_job(self.write_config_file)
-
-    def write_config_file(self):
```
Grepping through the code, I could find no current reference to code that actually _reads_ this outputted config file.

I looked through the history of this component and it appears this was previously part of the old python library this component used for the harmony api, prior to moving to `aioharmony`. I think in the process of moving libraries, the previous authors started having this entity write this out since the new library didn't contain that functionality.

Apologies if there is a common pattern I was missing here, but I could find no reason to continue writing this out. Removing this simplified writing more tests around the state handling of the entities in this integration..

bdraco replied:
The conf file is used by humans to figure out which buttons to put in their automations

mkeesey commented on homeassistant/components/harmony/remote.py:
```diff
@@ -405,94 +319,21 @@ async def async_send_command(self, command, **kwargs):
-                self._config_path,
-                exc,
-            )
+    async def sleep(self, time):
```
Had better luck splitting this into a specific method I could patch rather than patching `asyncio.sleep` directly in the connection state change tests. I was probably missing something subtle in how I was attempting to mock it away, so apologies if this is not a best practice.

mkeesey commented on tests/components/harmony/test_config_flow.py:
```diff
@@ -16,23 +16,6 @@ def _get_mock_harmonyapi(connect=None, close=None):
 
 
-def _get_mock_harmonyclient():
-    harmonyclient_mock = MagicMock()
```
I've transitioned most of this to an actual fake in conftest.py - this allows many more tests to be written that deal with state changes and state management.

mkeesey committed:
Fix linting

mkeesey committed:
test sending activity change commands by id

mkeesey commented:
Working to get the code coverage up.

mkeesey committed:
Improving coverage

mkeesey committed:
Testing channel change commands

mkeesey committed:
Splitting subscriber logic into it's own class

mkeesey committed:
Improve coverage and tighten up .coveragerc

mkeesey committed:
Test cleanups.

mkeesey commented:
Sorry for the long cycle on this, but I think this PR is in a reasonable place now to be reviewed.

There is quite a bit of coverage with the `data.py` and `remote.py` files now, except for the error handling cases that were lacking before. I think it'll take a bit more scaffolding to really test those cases and I think this PR has probably hung around long enough without boy-scouting all of it. That logic hasn't really changed, so I don't feel there is a lot of risk to it.

Apologies if there are still some stylistic issues with this - happy to address issues that you still see.

bdraco commented on homeassistant/components/harmony/subscriber.py:
```diff
@@ -0,0 +1,98 @@
+    @callback
+    def async_subscribe(self, update_callback: HarmonyCallback) -> Callable:
+        """Add a callback subscriber."""
+        self._subscriptions.append(update_callback)
```
Wrap these in `HassJob` and then call `hass.async_run_job` on the jobs in the array.

mkeesey replied:
I rejiggered this a little bit. The callbacks within this HarmonyCallback structure are optional, so not necessarily all of them will be present. Instead of doing logic twice to check for their existence (once while wrapping, and again when calling later on) or keeping several sets of callbacks I used `async_run_job` directly. I don't think the lack of memoization will really affect performance that much.

Thank you for suggesting this - this path is much more convenient compared to manually handling async vs standard function calls.

bdraco commented on homeassistant/components/harmony/subscriber.py:
```diff
@@ -0,0 +1,98 @@
+                else:
+                    current_callback()
+
+    def _activity_starting(self, activity_info: tuple) -> None:
+        _LOGGER.debug("activity %s starting", activity_info)
+        for subscription in self._subscriptions:
+            current_callback = subscription.activity_starting
+            if current_callback:
+                if asyncio.iscoroutinefunction(current_callback):
+                    asyncio.create_task(current_callback(activity_info))
+                else:
+                    current_callback(activity_info)
+
+    def _activity_started(self, activity_info: tuple) -> None:
+        _LOGGER.debug("activity %s started", activity_info)
+        for subscription in self._subscriptions:
+            current_callback = subscription.activity_started
+            if current_callback:
+                if asyncio.iscoroutinefunction(current_callback):
+                    asyncio.create_task(current_callback(activity_info))
+                else:
+                    current_callback(activity_info)
```
This code is almost the same. Can you make it more DRY?

mkeesey replied:
Definitely. I got a bit lazy here - things have now been consolidated into a common function `_call_callbacks`.

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -331,7 +282,7 @@ async def got_disconnected(self, _=None):
         self._available = False
         # We're going to wait for 10 seconds before announcing we're
         # unavailable, this to allow a reconnection to happen.
-        await asyncio.sleep(10)
+        await self.sleep(10)
```
Can we use `asyncio.Event` here instead so its more reliable?

mkeesey replied:
Sorry, I'm struggling a bit to understand your suggestion. I'm a bit newer to async/await style concurrency, so sorry if I'm missing something. :smile: I tried grepping around the repo quite a bit and it seems to be pretty common to just await on chains of async functions and should reliably wait until the `await asyncio.sleep` call completes within our `sleep` wrapper.

bdraco replied:
Let's do something like this instead and avoid the sleep:

```
TIME_MARK_DISCONNECTED = 10

```
```
self._unsub_mark_disconnected = async_call_later(self.hass, TIME_MARK_DISCONNECTED, self._mark_disconnected_if_unavailable)
```

Then call `self._unsub_mark_disconnected` if its not `None` when removing the entity.

mkeesey replied:
Makes a ton of sense - thank you! Should be updated.

mkeesey committed:
re-add config file writing for harmony remote

mkeesey committed:
Create fixture for the mock harmonyclient

mkeesey committed:
Reduce duplication in subscription callbacks

mkeesey committed:
use async_run_job to call callbacks

mkeesey committed:
Adding some tests for async behaviors with subscribers.

mkeesey committed:
async_call_later for delay in marking remote unavailable

mkeesey committed:
Test disconnection handling in harmony remote

bdraco commented:
I'll try to find some time to do testing tomorrow

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -325,17 +288,18 @@ async def got_connected(self, _=None):
-        if not self._available:
-            # Still disconnected. Let the state engine know.
-            self.async_write_ha_state()
+        self._unsub_mark_disconnected = async_call_later(
```
Should switches have the same behavior on disconnection?  Maybe make this functionality a base/mix-in class for both switches and remotes?

mkeesey replied:
Originally I thought it made more sense for the switches to respond immediately (while keeping the remote behavior the same). Thinking about it more, that's pretty silly. Anyhow, I've split out a mixin to handle this and updated the remote slightly to get the availability state consistently with how the switches do it.

It turns out that the remote had to set it's availability state in a bunch of places because (even in the old code) it would subscribe to callbacks _after_ connection callbacks would be called the first time around. The data object we're using configures these callbacks prior to making the first connection so this is no longer an issue.

bdraco commented:
Tested and working as expected.  Just a few cleanups mentioned above remaining.

mkeesey committed:
Early exit if activity not specified

mkeesey committed:
Use connection state mixin

mkeesey committed:
Lint fix after rebase

mkeesey committed:
Fix isort

mkeesey commented:
I needed to rebase to catch up with some test changes upstream (to remove async_mock). The new commits are the newest 5, starting with 9861f2b346a4ae38da9e107a7849e2f0ed99feb9.

bdraco commented on homeassistant/components/harmony/connection_state.py:
```diff
@@ -0,0 +1,43 @@
+
+    def __init__(self):
+        """Initialize this mixin instance."""
+        self._unsub_mark_disconnected = None
```
```suggestion
        super().__init__()
        self._unsub_mark_disconnected = None
```

This may need a super init.  Need to check

mkeesey replied:
Good catch - it seemed to work without but best to have it be consistent.

mkeesey committed:
super init for ConnectionStateMixin

bdraco commented:
I'll try to run though this in the afternoon (HST)

bdraco commented:
@mkeesey 

Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

mkeesey committed:
Adding @mkeesey to harmony CODEOWNERS

mkeesey commented:
@bdraco I should be added to CODEOWNERS - thanks for all your help on this review!

bdraco reviewed (approved):
Everything appears to be working as executed 👍 

I'll wait a day before merging this to do some more testing.

If another reviewer has followup comments, please address them in a new PR if they come after merge.

bdraco commented:
I didn't notice it until I was testing #45964, but the unique id for the activity isn't stable.

If the activity gets renamed, it will change.

We should use the activity id (check the harmony_*.conf) for it (api likely has a way to get it).

We will need to migrate the existing ones with a unique id migrator (example in powerwall)

mkeesey commented:
@bdraco makes sense! I'll try to take a crack at it sometime in the next few days.

MartinHjelmare commented on tests/components/harmony/test_commands.py:
```diff
@@ -0,0 +1,263 @@
+"""Test sending commands to the Harmony Hub remote."""
```
Our standard is to divide the tests in modules that correspond to the platforms the integration supports plus config flow and init module tests. Ie the switch platform should have tests in a module named `test_switch.py` etc.

MartinHjelmare commented on tests/components/harmony/test_connection_changes.py:
```diff
@@ -0,0 +1,67 @@
+    await hass.config_entries.async_setup(entry.entry_id)
+    await hass.async_block_till_done()
+
+    data = hass.data[DOMAIN][entry.entry_id]
```
Please don't access integration details in the tests. Use the core interfaces to start the test and patch in the periphery, eg the library, to get access to callbacks and mock library state.

mkeesey commented:
I'll take a crack at the test refactoring in a subsequent PR to keep each of them a bit smaller (and easier to review for folks).
